### Why Data Cleaning Matters
  A machine learning model is only as good as the data it learns from.
  If we feed it messy data — returns mixed with purchases, missing customer
  IDs, cancelled orders — it will learn the wrong patterns and make
  unreliable predictions.

  Think of it this way: if you were trying to understand a customer's
  buying behavior but half their "purchases" were actually returns,
  you would completely misread who they are as a customer.

  ### Our Cleaning Strategy
  Based on what we found in Chapter 1, we need to:
  1. Remove rows with no CustomerID — we cannot track anonymous guests
  2. Remove cancelled orders — InvoiceNo starting with 'C'
  3. Remove negative Quantity and UnitPrice — these are returns
  4. Drop missing descriptions — minor cleanup
  5. Create TotalAmount = Quantity × UnitPrice — our core revenue metric

In [2]:
import pandas as pd

In [3]:
df = pd.read_excel('/Users/purvamugdiya/Desktop/customer-intelligence-platform/data/Online Retail.xlsx')

In [ ]:
  # 1. Drop missing CustomerIDs
df = df.dropna(subset=['CustomerID'])

In [ ]:
  # 3. Remove negative Quantity and UnitPrice
df = df[df['Quantity'] > 0]
df = df[df['UnitPrice'] > 0]

In [ ]:
  # 4. Drop missing descriptions
df = df.dropna(subset=['Description'])

In [8]:
 # 5. Convert CustomerID to integer
df['CustomerID'] = df['CustomerID'].astype(int)

In [9]:
  # 6. Create TotalAmount column
df['TotalAmount'] = df['Quantity'] * df['UnitPrice']

In [12]:
df.to_csv('../data/cleaned_retail.csv', index=False)
print("Saved! Columns:", df.columns.tolist())

Saved! Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country', 'TotalAmount']


In [10]:
# Check results
print("Shape after cleaning:", df.shape)
print("\nMissing values:\n", df.isnull().sum())
print("\nSample:")
df.head()

Shape after cleaning: (397884, 9)

Missing values:
 InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
TotalAmount    0
dtype: int64

Sample:


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalAmount
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34


### Results — What Changed?

  We went from 541,909 rows to 397,884 rows — removing 144,025 records.

  Here is what each cleaning step removed and why it matters:

  | Step | Records Removed | Reason |
  |---|---|---|
  | Missing CustomerID | ~135,080 | Cannot track anonymous customers |
  | Cancelled orders | ~9,288 | Not real purchases |
  | Negative values | ~10,624 | Returns and data errors |
  | Missing descriptions | ~1,454 | Minor data quality issue |

  ### So What?
  The 397,884 remaining rows represent clean, verified purchases by
  identifiable customers. Every row now tells a reliable story about
  a real customer making a real purchase — which is exactly what our
  models need to learn from.

  The TotalAmount column we created (Quantity × UnitPrice) gives us
  a single reliable number for the value of each transaction — the
  foundation of our revenue and monetary analysis.